x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRestoreSourceService.java

```
@@ -154,22 +191,70 @@ private RestoreContext(String sessionUUID, IndexShard indexShard, Engine.IndexCo
-            assert Thread.holdsLock(CcrRestoreSourceService.this);
-            removeSessionForShard(sessionUUID, indexShard);
-            IOUtils.closeWhileHandlingException(commitRef);
-        }
+        private long readFileBytes(String fileName, BytesReference reference) throws IOException {
+            Releasable lock = keyedLock.tryAcquire(fileName);
+            if (lock == null) {
+                throw new IllegalStateException("can't read from the same file on the same session concurrently");
+            }
+            try (Releasable releasable = lock) {
+                final IndexInput indexInput = cachedInputs.computeIfAbsent(fileName, f -> {
+                    try {
+                        return commitRef.getIndexCommit().getDirectory().openInput(fileName, IOContext.READONCE);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                });
 
-        private void removeSessionForShard(String sessionUUID, IndexShard indexShard) {
-            logger.debug("closing session [{}] for shard [{}]", sessionUUID, indexShard.shardId());
-            HashSet<String> sessions = sessionsForShard.get(indexShard);
-            if (sessions != null) {
-                sessions.remove(sessionUUID);
-                if (sessions.isEmpty()) {
-                    sessionsForShard.remove(indexShard);
+                BytesRefIterator refIterator = reference.iterator();
+                BytesRef ref;
+                while ((ref = refIterator.next()) != null) {
+                    byte[] refBytes = ref.bytes;
+                    indexInput.readBytes(refBytes, 0, refBytes.length);
                 }
+
+                long offsetAfterRead = indexInput.getFilePointer();
+
+                if (offsetAfterRead == indexInput.length()) {
+                    cachedInputs.remove(fileName);
+                    IOUtils.closeWhileHandlingException(indexInput);
```

lets not swallow exceptions here. just close it.